[macOS] Send playEffect() twice to Gamepad.actuator and it cannot stop with reset()

basuke opened the pull request:
#### 467d7dd6f1ea3e0b36eb4e3c01869663b066831a
<pre>
[macOS] Send playEffect() twice to Gamepad.actuator and it cannot stop with reset()
<a href="https://bugs.webkit.org/show_bug.cgi?id=279094">https://bugs.webkit.org/show_bug.cgi?id=279094</a>
<a href="https://rdar.apple.com/126589062">rdar://126589062</a>

Reviewed by Chris Dumez.

When sending playEffect() to Gamepad.actuator more than once, the reset() cannot stop the gamepad from vibrating.

The root cause of the issue is requesting a completionHandler invocation to wrong effect object.
Say two sequential playEffect requests is coming.

1. The effect object of request 1 is stored in m_currentDualRumbleEffect and referenced by currentEffect in playEffect() method.
2. In second playEffect(), currentEffect is stopped  before start playing second one. This is in main thread.
3. Notification for stopPlaying is called for first request {*) in CoreHaptics thread.
4. currentEffect is replaced with new effect.
5. Notification arrived in main thread and try to invoke completionHandler targeting to second request. This is wrong because the notification is for first request.

Move the play logic from GameControllerHapticEngines to GameControllerHapticEffect to avoid this confusion.

* Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEffect.h:
* Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEffect.mm:
(WebCore::GameControllerHapticEffect::create):
(WebCore::GameControllerHapticEffect::GameControllerHapticEffect):
(WebCore::GameControllerHapticEffect::~GameControllerHapticEffect):
(WebCore::GameControllerHapticEffect::start):
(WebCore::GameControllerHapticEffect::ensureStarted):
(WebCore::GameControllerHapticEffect::startEngine):
(WebCore::GameControllerHapticEffect::registerNotification):
(WebCore::GameControllerHapticEffect::stop):
(WebCore::GameControllerHapticEffect::leftEffectFinishedPlaying): Deleted.
(WebCore::GameControllerHapticEffect::rightEffectFinishedPlaying): Deleted.
* Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEngines.h:
* Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEngines.mm:
(WebCore::GameControllerHapticEngines::create):
(WebCore::GameControllerHapticEngines::playEffect):
(WebCore::GameControllerHapticEngines::ensureStarted): Deleted.

Canonical link: <a href="https://commits.webkit.org/286251@main">https://commits.webkit.org/286251@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a436b64eff4b78881ba7c2145346bf99b536c97e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75325 "Passed style check") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28165 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/79796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/26590 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77442 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63909 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2549 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/79796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/26590 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78392 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/49306 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64709 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/79796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/46705 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22202 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/24917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/67727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22541 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/81282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/2655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/1676 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/81282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/2816 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64705 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/81282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16581 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/8763 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/2622 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/5443 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/2647 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/3582 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/2654 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEffect.h:
```diff
@@ -56,12 +58,14 @@ class GameControllerHapticEffect : public CanMakeWeakPtr<GameControllerHapticEff
-private:
-    GameControllerHapticEffect(RetainPtr<id>&& leftPlayer, RetainPtr<id>&& rightPlayer);
-
+    CHHapticEngine* m_leftEngine;
```
Can this be a smart pointer?

cdumez commented on Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEffect.h:
```diff
@@ -56,12 +58,14 @@ class GameControllerHapticEffect : public CanMakeWeakPtr<GameControllerHapticEff
-    GameControllerHapticEffect(RetainPtr<id>&& leftPlayer, RetainPtr<id>&& rightPlayer);
-
+    CHHapticEngine* m_leftEngine;
+    CHHapticEngine* m_rightEngine;
```
ditto.

cdumez commented on Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEffect.h:
```diff
@@ -56,12 +58,14 @@ class GameControllerHapticEffect : public CanMakeWeakPtr<GameControllerHapticEff
     void stop();
 
+private:
+    GameControllerHapticEffect(CHHapticEngine* leftEngine, CHHapticEngine* rightEngine, RetainPtr<id>&& leftPlayer, RetainPtr<id>&& rightPlayer);
```
For Objective C objects, we put the star on the other side.

cdumez commented on Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEffect.mm:
```diff
@@ -79,30 +80,40 @@ static double magnitudeToIntensity(double magnitude)
         return retainPtr([engine createPlayerWithPattern:pattern.get() error:&error]);
     };
 
+    CHHapticEngine* leftEngine;
```
Smart pointer?

cdumez commented on Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEffect.mm:
```diff
@@ -79,30 +80,40 @@ static double magnitudeToIntensity(double magnitude)
     };
 
+    CHHapticEngine* leftEngine;
+    CHHapticEngine* rightEngine;
```
Smart pointer?

cdumez commented on Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEffect.mm:
```diff
@@ -79,30 +80,40 @@ static double magnitudeToIntensity(double magnitude)
+        rightPlayer = createPlayer(rightEngine, parameters.rightTrigger);
         break;
         }
+    default:
```
Do we need a default case?

cdumez commented on Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEffect.mm:
```diff
@@ -98,11 +105,13 @@ static double magnitudeToIntensity(double magnitude)
 
-GameControllerHapticEffect::GameControllerHapticEffect(RetainPtr<id>&& leftPlayer, RetainPtr<id>&& rightPlayer)
-    : m_leftPlayer(WTFMove(leftPlayer))
+GameControllerHapticEffect::GameControllerHapticEffect(RetainPtr<CHHapticEngine> leftEngine, RetainPtr<CHHapticEngine> rightEngine, RetainPtr<id>&& leftPlayer, RetainPtr<id>&& rightPlayer)
```
`RetainPtr<CHHapticEngine>&&`

cdumez reviewed (approved):
LGTM with nit fix.

cdumez commented on Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEffect.mm:
```diff
@@ -127,8 +159,44 @@ static double magnitudeToIntensity(double magnitude)
+            return CHHapticEngineFinishedActionStopEngine;
+        }).get()];
+#else
+        RunLoop::main().dispatchAfter(m_delay + m_duration, [weakThis = WeakPtr { *this }] {
```
This doesn't seem like something we should ship. It looks like a workaround for a bug we haven't figured out yet.

cdumez commented on Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEffect.h:
```diff
@@ -43,19 +45,24 @@ class GameControllerHapticEffect : public RefCountedAndCanMakeWeakPtr<GameContro
 public:
     static RefPtr<GameControllerHapticEffect> create(GameControllerHapticEngines&, GamepadHapticEffectType, const GamepadEffectParameters&);
-    ~GameControllerHapticEffect();
+    virtual ~GameControllerHapticEffect();
```
Why is this virtual now?

basuke committed:
[macOS] Send playEffect() twice to Gamepad.actuator and it cannot stop with reset()
https://bugs.webkit.org/show_bug.cgi?id=279094
rdar://126589062

Reviewed by Chris Dumez.

When sending playEffect() to Gamepad.actuator more than once, the reset() cannot stop the gamepad from vibrating.

The root cause of the issue is requesting a completionHandler invocation to wrong effect object.
Say two sequential playEffect requests is coming.

1. The effect object of request 1 is stored in m_currentDualRumbleEffect and referenced by currentEffect in playEffect() method.
2. In second playEffect(), currentEffect is stopped  before start playing second one. This is in main thread.
3. Notification for stopPlaying is called for first request {*) in CoreHaptics thread.
4. currentEffect is replaced with new effect.
5. Notification arrived in main thread and try to invoke completionHandler targeting to second request. This is wrong because the notification is for first request.

Move the play logic from GameControllerHapticEngines to GameControllerHapticEffect to avoid this confusion.

* Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEffect.h:
* Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEffect.mm:
(WebCore::GameControllerHapticEffect::create):
(WebCore::GameControllerHapticEffect::GameControllerHapticEffect):
(WebCore::GameControllerHapticEffect::~GameControllerHapticEffect):
(WebCore::GameControllerHapticEffect::start):
(WebCore::GameControllerHapticEffect::ensureStarted):
(WebCore::GameControllerHapticEffect::startEngine):
(WebCore::GameControllerHapticEffect::registerNotification):
(WebCore::GameControllerHapticEffect::stop):
(WebCore::GameControllerHapticEffect::leftEffectFinishedPlaying): Deleted.
(WebCore::GameControllerHapticEffect::rightEffectFinishedPlaying): Deleted.
* Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEngines.h:
* Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEngines.mm:
(WebCore::GameControllerHapticEngines::create):
(WebCore::GameControllerHapticEngines::playEffect):
(WebCore::GameControllerHapticEngines::ensureStarted): Deleted.

Canonical link: https://commits.webkit.org/286251@main
